driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionExternalKeyVaultTest.java

```diff
@@ -0,0 +1,187 @@
+        }
+
+        AutoEncryptionSettings autoEncryptionSettings = autoEncryptionSettingsBuilder.build();
+
+        MongoClientSettings clientSettings = clientSettingsBuilder
+                .autoEncryptionSettings(autoEncryptionSettings)
+                .build();
+        clientEncrypted = MongoClients.create(clientSettings);
+
+        ClientEncryptionSettings.Builder clientEncryptionSettingsBuilder = ClientEncryptionSettings.builder().
+                keyVaultMongoClientSettings(clientSettingsBuilder.build())
+                .kmsProviders(kmsProviders)
+                .keyVaultNamespace("admin.datakeys");
+
+        if (withExternalKeyVault) {
+            clientEncryptionSettingsBuilder.keyVaultMongoClientSettings(externalClientSettings);
+        }
+
+        ClientEncryptionSettings clientEncryptionSettings = clientEncryptionSettingsBuilder.build();
+        clientEncryption = ClientEncryptions.create(clientEncryptionSettings);
+    }
+
+    @Test
+    public void testExternal() {
+        boolean authExceptionThrown = false;
+        MongoCollection<BsonDocument> coll = clientEncrypted
+                .getDatabase("db")
+                .getCollection("coll", BsonDocument.class);
+        FutureResultCallback<Void> voidCallback;
+        try {
+            voidCallback = new FutureResultCallback<Void>();
+            coll.insertOne(new BsonDocument().append("encrypted", new BsonString("test")), voidCallback);
+            voidCallback.get();
+        } catch (MongoSecurityException mse) {
+            authExceptionThrown = true;
+        }
+        assertEquals(authExceptionThrown, withExternalKeyVault);
+
+        EncryptOptions encryptOptions = new EncryptOptions("AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic")
+                .keyId(new BsonBinary(BsonBinarySubType.UUID_STANDARD, Base64.getDecoder().decode("LOCALAAAAAAAAAAAAAAAAA==")));
+        authExceptionThrown = false;
+        try {
+            FutureResultCallback<BsonBinary> bsonBinaryCallback = new FutureResultCallback<BsonBinary>();
+            clientEncryption.encrypt(new BsonString("test"), encryptOptions, bsonBinaryCallback);
+            bsonBinaryCallback.get();
+        } catch (MongoSecurityException mse) {
+            authExceptionThrown = true;
+        }
+        assertEquals(authExceptionThrown, withExternalKeyVault);
+    }
+
+    private static BsonDocument bsonDocumentFromPath(final String path) throws IOException, URISyntaxException {
+        return getTestDocument(new File(ClientSideEncryptionExternalKeyVaultTest.class
+                .getResource("/client-side-encryption-external/" + path).toURI()));
+    }
+
+    @Parameterized.Parameters(name = "withExternalKeyVault: {0}")
+    public static Collection<Object[]> data() {
+        return Arrays.asList(new Object[]{true}, new Object[]{false});
+    }
+
+    @After
+    public void after() {
+        if (clientEncrypted != null) {
+            try {
```

Comment:
probably we need to create a ticket to investigate why we have `failed` closing sometimes.